### PR TITLE
Enable GPU access with DeviceRequests

### PR DESCRIPTION
**Pull request by aiordache**
Convert compose-spec `devices` mapping to `DeviceRequest` to enable GPU access to containers.

Tested on a GPU host:

```
$ cat /etc/docker/daemon.json 
{
    "runtimes": {
        "nvidia": {
            "path": "nvidia-container-runtime",
            "runtimeArgs": []
        }
    }
}
```
Sample compose files:
```yaml
services:
  test:
    image: nvidia/cuda
    command: nvidia-smi
    runtime: nvidia
```
or 
```yaml
services:
  test:
    image: nvidia/cuda
    command: nvidia-smi
    deploy:
      resources:
        reservations:
          devices:
          - 'driver': 'nvidia'
            'count': 1
            'capabilities': ['gpu', 'utility']
```

```
$ docker-compose up
Creating network "gpu_default" with the default driver
Creating gpu_test_1 ... done
Attaching to gpu_test_1
test_1  | Fri Nov 13 20:46:11 2020       
test_1  | +-----------------------------------------------------------------------------+
test_1  | | NVIDIA-SMI 450.80.02    Driver Version: 450.80.02    CUDA Version: 11.1     |
test_1  | |-------------------------------+----------------------+----------------------+
test_1  | | GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
test_1  | | Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
test_1  | |                               |                      |               MIG M. |
test_1  | |===============================+======================+======================|
test_1  | |   0  Tesla T4            On   | 00000000:00:1E.0 Off |                    0 |
test_1  | | N/A   23C    P8     9W /  70W |      0MiB / 15109MiB |      0%      Default |
test_1  | |                               |                      |                  N/A |
test_1  | +-------------------------------+----------------------+----------------------+
test_1  |                                                                                
test_1  | +-----------------------------------------------------------------------------+
test_1  | | Processes:                                                                  |
test_1  | |  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
test_1  | |        ID   ID                                                   Usage      |
test_1  | |=============================================================================|
test_1  | |  No running processes found                                                 |
test_1  | +-----------------------------------------------------------------------------+
gpu_test_1 exited with code 0

```

```yaml
services:
  test:
    image: tensorflow/tensorflow:latest-gpu
    command: python -c "import tensorflow as tf;print(tf.test.gpu_device_name())"
    deploy:
      resources:
        reservations:
          devices:
          - 'driver': 'nvidia'
            'capabilities': ['gpu']
```
```
$ docker-compose up
Creating network "gpu_default" with the default driver
Creating gpu_test_1 ... done
Attaching to gpu_test_1
test_1  | 2020-11-13 20:49:54.444634: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
.....
test_1  | 2020-11-13 20:49:56.048674: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1402] Created TensorFlow device (/device:GPU:0 with 13970 MB memory) -> physical GPU (device: 0, name: Tesla T4, pci bus id: 0000:00:1e.0, compute capability: 7.5)
test_1  | /device:GPU:0
gpu_test_1 exited with code 0

```
Tested on a multi-GPU host:

```
$ nvidia-smi 
Fri Nov 13 20:57:48 2020       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 450.80.02    Driver Version: 450.80.02    CUDA Version: 11.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla T4            On   | 00000000:00:1B.0 Off |                    0 |
| N/A   72C    P8    12W /  70W |      0MiB / 15109MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  Tesla T4            On   | 00000000:00:1C.0 Off |                    0 |
| N/A   67C    P8    11W /  70W |      0MiB / 15109MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   2  Tesla T4            On   | 00000000:00:1D.0 Off |                    0 |
| N/A   74C    P8    12W /  70W |      0MiB / 15109MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   3  Tesla T4            On   | 00000000:00:1E.0 Off |                    0 |
| N/A   62C    P8    11W /  70W |      0MiB / 15109MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

```

Enable access only to GPU-0 and GPU-3 devices
```yaml
services:
  test:
    image: tensorflow/tensorflow:latest-gpu
    command: python -c "import tensorflow as tf;print(tf.test.gpu_device_name())"
    deploy:
      resources:
        reservations:
          devices:
          - 'driver': 'nvidia'
            'device_ids': ['0','3']
            'capabilities': ['gpu']
```

```
$ docker-compose up
...
test_1  | 2020-11-13 21:02:52.076151: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1402] Created TensorFlow device (/device:GPU:0 with 13970 MB memory) -> physical GPU (device: 0, name: Tesla T4, pci bus id: 0000:00:1b.0, compute capability: 7.5)
test_1  | 2020-11-13 21:02:52.076752: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:982] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
test_1  | 2020-11-13 21:02:52.077844: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1402] Created TensorFlow device (/device:GPU:1 with 13970 MB memory) -> physical GPU (device: 1, name: Tesla T4, pci bus id: 0000:00:1e.0, compute capability: 7.5)
test_1  | /device:GPU:0
gpu_test_1 exited with code 0
```

Requires https://github.com/compose-spec/compose-spec/pull/109
Closes https://github.com/docker/compose/issues/6691

### CODE REVIEW

**Comment by opptimus** [comment]
Wonderful and significant features, expect to be compiled！

File: compose/service.py [hunk 1]
```diff
@@ -179,6 +180,7 @@ def __init__(
             ipc_mode=None,
             pid_mode=None,
             default_platform=None,
+            device_requests=None,
```
Comment (chris-crone) [inline compose/service.py:183]: Careful of changing function parameter ordering. It's possible that users are calling this function with positional parameters so it's best to add new parameters to the end of the list (i.e.: after `extra_labels`).

**Review by chris-crone — APPROVED** [review]
One minor change but LGTM

**Review by gtardif — APPROVED** [review]

**Comment by rowandempster** [comment]
Hey @aiordache, (let me know if I should move this comment over to an issue)

I'm trying out this change using the prerelease `docker-compose version 1.28.0-rc1, build 1f469474` in order to launch CarlaSim with access to my server's GPUs, but to no avail :(

Using `docker run`:
```
docker run -it -e --gpus all carlasim/carla:0.9.10.1 ./CarlaUE4.sh Town02 -quality-level=Low -carla-server
```
Carla runs successfully

Using `docker-compose`:
```
carla_server:
    image: carlasim/carla:0.9.10.1
    runtime: nvidia
    command: ./CarlaUE4.sh Town02 -quality-level=Low -carla-server
``` 
Carla crashes in the same way it does when I exclude the `--gpus all` flag from `docker run`.

**Pull request merged**